### PR TITLE
specify TPI version

### PR DESCRIPTION
**Pull request by casperdcl**
- Specify minimum terraform-provider-iterative version (may fix edge cases)
- follow-up to #883

### CODE REVIEW

**Comment by 0x2b3bfa0** [comment]
🤔 What edge cases is this change trying to cover?

**Comment by casperdcl** [comment]
Things related to https://github.com/iterative/terraform-provider-iterative/pull/266?

**Comment by DavidGOrtega** [comment]
We do not tie the latest version of terraform. Indeed we propagate a non versioned terraform state so the terraform installed in the cloud instance is able to run the state.

Just to clarify. We are not tied a a specific terraform version

**Comment by dacbd** [comment]
could be nice, in combo with: https://github.com/iterative/terraform-provider-iterative/pull/255 for users to have the option to lock their automation into a tested/proven release (for them).


For example, I had a failure, and at least one other user from discord where a new release of tpi was out that cml tried to use but the registry hadn't quite propagated completely and jobs failed, granted this was an edge case of some stars aligning just right with when a job was running / when a release was made.


Yes if not set, default to latest 💯

**Comment by dacbd** [comment]
> But this PR the thing it does it to enforce that the tpi version is above a minimum and not to specify a specific terraform version, right? @dacbd
> 
> ```
> iterative = { source = "iterative/iterative", version = ">= 0.9.10" }
> ```

I confess to only reading the comments 🐍

**Comment by dacbd** [comment]
my brain auto translated "specify TPI version" to "Allow user to specify TPI version"

**Comment by casperdcl** [comment]
I think at least letting the user specify the TPI version as @dacbd suggests is a good idea.

- TPI may introduce a breaking change in future
- there may be some time for a change to propagate to a CML patch release (e.g. #883)
- right now, if a user has already manually run `terraform apply` there may be an old cached version of TPI which CML will use, right?

**Comment by DavidGOrtega** [comment]
> I think at least letting the user specify the TPI version as @dacbd suggests is a good idea.

Just for the record CML unversion the tfstate from terraform. This is done to avoid minor terraform changes that avoids the run. 
Historically we have been always backwards compatible. Indeed `task` could be the first time that we are changing behaviour from version to version (however we are in alpha mode so probably in the future we will stick to the plan)

**Comment by dacbd** [comment]
> Just for the record CML unversion the tfstate from terraform.

@DavidGOrtega are you referring to when cml is invoked with base64 tfstate on the server?

**Comment by DavidGOrtega** [comment]
> @DavidGOrtega are you referring to when cml is invoked with base64 tfstate on the server?

Yes, one of the biggest issues was having synced the terraform version in the remote host and the user's local one. If the differ terraform will fail. This topic was a hell back in the day for terraform users that discovered unpleasantly that they were not able to terminate instances because they updated terraform

**Comment by DavidGOrtega** [comment]
In any case. Does this PR solve the problem? Should we merge it? Should we close it? @iterative/cml

**Comment by dacbd** [comment]
I'm happy to add the selection feature I want to see as a new PR or as one targeting this branch (convert this PR to draft for now?)

**Comment by dacbd** [comment]
@iterative/cml this half works. the invoking cml can have its tpi specified via passing a [version constraint]( https://www.terraform.io/language/expressions/version-constraints) with `--tpi-version >= 0.9.10` / `--tpi-version =0.10.3` but the instances installed version of the provider is "latest" I haven't investigated changing that yet, but my hunch is it would > non-trivial

I feel this doesn't quite match the "expected behavior" but on the other hand the instance's tpi only preforms the delete action and none of the setup/configuration... 👍 👎

**Comment by 0x2b3bfa0** [comment]
> _I haven't investigated changing that yet, but my hunch is it would > non-trivial:

From what I understand, something along the lines of https://github.com/iterative/terraform-provider-iterative/pull/103 should suffice to close the [stange loop](https://en.wikipedia.org/wiki/Strange_loop).

**Comment by dacbd** [comment]
Yeah, that was my initial thought about how to get it there as well but feels extra redundant, you pick your version or constraint which is part of the provider definition but also pass in the resource definition?

I would want to try embedding it [here](https://github.com/iterative/terraform-provider-iterative/blob/master/iterative/resource_runner.go#L399) so that it is communicated to the instance via base64 tfstate. /my "haven't investigated" part

**Comment by dacbd** [comment]
> Yeah, that was my initial thought about how to get it there as well but feels extra redundant, you pick your version or constraint which is part of the provider definition but also pass in the resource definition?
> 
> I would want to try embedding it [here](https://github.com/iterative/terraform-provider-iterative/blob/master/iterative/resource_runner.go#L399) so that it is communicated to the instance via base64 tfstate. /my "haven't investigated" part

being able to parse it out of the tfstate like that would make it a simple adjustment here:
https://github.com/iterative/cml/blob/01b2cef37b3efaf2241d0ee945ccc71662a9c3d2/bin/cml/runner.js#L375-L383
as a param to line 375

**Comment by 0x2b3bfa0** [comment]
Not sure if it's going to work, but definitely worth a try. Good initiation in the path of the Terraform Heresy. 🎉 😄

**Review by 0x2b3bfa0 — APPROVED** [review]
Static review: looks good to me!

**Comment by dacbd** [comment]
ahh yes, tested via:
```
bin/cml.js runner \
    --single \
    --tpi-version =0.10.3 \
    --labels=tpi_version \
    --token=*** \
    --cloud=aws \
    --cloud-region=us-west-2 \
    --cloud-type=t3.small
```

Cat of the generated main.tf / lock file

```
provider "registry.terraform.io/iterative/iterative" {
  version     = "0.10.3"
  constraints = "0.10.3"
  hashes = [
    "h1:qG6mYJ7/v/RvSvAefma7xdvrzxViyIvP6K35IjvlPU0=",
    "zh:33e6f437aeae3834f5ac34c5b39982dcd118169c039232d327e5dae17faa4283",
    "zh:8f72ca6dad3fdffbb7a5f267ac8d1d0d452a4e31e8764a54aa17d55fcf126637",
    "zh:975b4511e2592e46e532b6d2e9a7d57763592742f968ad3100021ccff531102b",
    "zh:a634f4d77634b5cc5f0599a3688b8bc363d9f2889086671d6f07a57421c7d15d",
    "zh:b4b648188a85c77c7285247d9830cf6ce0ac4c67[41](https://github.com/dacbd/cml-pulse/runs/6010390374?check_suite_focus=true#step:6:41)5dc4e6b0794f3d7c950ad7",
    "zh:cc5624f08b9b1e[45](https://github.com/dacbd/cml-pulse/runs/6010390374?check_suite_focus=true#step:6:45)14fac5b324c150713c2b9e63c7f3874f37792e125ca72aed",
    "zh:e0707971f3cf1c0e011290e6e76a05b03183f756c91b5b9bd9c5c5027c8b2353",
    "zh:e5bb7ad0fae6d50ca04109cfaa57ef[47](https://github.com/dacbd/cml-pulse/runs/6010390374?check_suite_focus=true#step:6:47)a263fd6eb00eec[51](https://github.com/dacbd/cml-pulse/runs/6010390374?check_suite_focus=true#step:6:51)[61](https://github.com/dacbd/cml-pulse/runs/6010390374?check_suite_focus=true#step:6:61)0[62](https://github.com/dacbd/cml-pulse/runs/6010390374?check_suite_focus=true#step:6:62)3fecbb06c5e",
  ]
}
terraform {
  required_providers {
    iterative = { source = "iterative/iterative", version = "=0.10.3" }
  }
}
provider "iterative" {}

resource "iterative_cml_runner" "runner" {
  repo = "https://github.com/dacbd/cml-pulse"
  token = "***"
  driver = "github"
  labels = "tpi_version"
  cml_version = "0.14.1"
  idle_timeout = 300
  name = "cml-gjhlm6lw9j"
  single = "true"
  cloud = "aws"
  region = "us-west-2"
  instance_type = "t3.small"
  
  
  
  
  
  spot_price = -1
  
  
  
  metadata = {
    
  }
  docker_volumes = []
}
```

File: bin/cml/runner.js [hunk 1]
```diff
@@ -436,6 +439,12 @@ exports.handler = async (opts) => {
 exports.builder = (yargs) =>
   yargs.env('CML_RUNNER').options(
     kebabcaseKeys({
+      tpiVersion: {
```
Comment (casperdcl) [inline bin/cml/runner.js:442]: should this be `hidden: true`?

Comment (0x2b3bfa0) [reply bin/cml/runner.js:442]: Good point! 🤦🏼‍♂️ `cmlVersion` that is slightly less arcane is also hidden, so it makes sense ho hide both.

Comment (0x2b3bfa0) [reply bin/cml/runner.js:442]: * https://github.com/iterative/cml/pull/967

**Pull request merged**